### PR TITLE
add Openbsd support

### PR DESCRIPTION
**Pull request by goneri**
Based on: https://github.com/canonical/cloud-init/pull/62

### CODE REVIEW

File: cloudinit/distros/bsd_util.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# This file is part of cloud-init. See LICENSE file for license information.
+
```
Comment (igalic) [inline cloudinit/distros/bsd_util.py]: if this is based on #62, why is this file added again?

Comment (goneri) [reply cloudinit/distros/bsd_util.py]: because the NetBSD patch is in the first in the series :-).

File: cloudinit/distros/bsd_util.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            k, v = line.split('=', 1)
+            contents[k] = v
+    return contents.get(key)
+
+
+def set_rc_config_value(key, value, fn='/etc/rc.conf'):
+    lines = []
+    done = False
+    value = shlex.quote(value)
+    for line in util.load_file(fn).splitlines():
```
Comment (raharper) [inline cloudinit/distros/bsd_util.py]: Does util.load_shell_content() work for you here? or  are there non-shell spaces that break parsing?

File: cloudinit/distros/bsd_util.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            k, v = line.split('=', 1)
+            if k == key:
+                v = value
+                done = True
+            lines.append('='.join([k, v]))
+        else:
+            lines.append(line)
+    if not done:
+        lines.append('='.join([key, value]))
+    with open(fn, 'w') as fd:
```
Comment (raharper) [inline cloudinit/distros/bsd_util.py]: util.write_file()

File: cloudinit/distros/netbsd.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
+# Copyright (C) 2019-2020 Gonéri Le Bouder
+#
+# This file is part of cloud-init. See LICENSE file for license information.
+
+import crypt
+import os
+import platform
+import six
```
Comment (raharper) [inline cloudinit/distros/netbsd.py]: As we're removing six ... what to do about this?  @OddBloke ?

File: cloudinit/distros/netbsd.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        distros.Distro.__init__(self, name, cfg, paths)
+        # This will be used to restrict certain
+        # calls from repeatly happening (when they
+        # should only happen say once per instance...)
+        self._runner = helpers.Runners(paths)
+        self.osfamily = 'netbsd'
+        cfg['ssh_svcname'] = 'sshd'
+
+    def _read_system_hostname(self):
+        sys_hostname = self._read_hostname(filename='/etc/rc.conf')
```
Comment (raharper) [inline cloudinit/distros/netbsd.py]: There are many strings that are all '/etc/rc.conf'; you already have this as self.hostname_conf_fn , but maybe just have a self.rc_conf = '/etc/rc.conf' and use that instead.

File: cloudinit/distros/netbsd.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+        return ('/etc/rc.conf', sys_hostname)
+
+    def _read_hostname(self, filename, default=None):
+        return bsd_util.get_rc_config_value('hostname')
+
+    def _write_hostname(self, hostname, filename):
+        bsd_util.set_rc_config_value('hostname', hostname,
+                                     fn='/etc/rc.conf')
+
+    def create_group(self, name, members):
```
Comment (raharper) [inline cloudinit/distros/netbsd.py]: This looks a lot like the base class Distro.create_group;   what's different in here?  It appears to be the usermod command; so maybe in the base create_group you could make use of util.is_OpenBSD() ?

File: cloudinit/distros/netbsd.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+        adduser_cmd = ['useradd']
+        log_adduser_cmd = ['useradd']
+
+        adduser_opts = {
+            "homedir": '-d',
+            "gecos": '-c',
+            "primary_group": '-g',
+            "groups": '-G',
+            "shell": '-s',
```
Comment (raharper) [inline cloudinit/distros/netbsd.py]: This is a subset of user opts that are allowed in cloud-init user config;  if OpenBSD doesn;t support some of them, then here we should log which options provided are not getting set.

File: cloudinit/distros/netbsd.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+            method = crypt.METHOD_BLOWFISH  # pylint: disable=E1101
+            hashed_pw = crypt.crypt(
+                    password,
+                    crypt.mksalt(method))
+
+        try:
+            util.subp(['usermod', '-C', 'no', '-p', hashed_pw, user])
+        except Exception:
+            util.logexc(LOG, "Failed to set password for %s", user)
+            raise
```
Comment (raharper) [inline cloudinit/distros/netbsd.py]: If you have a crypt error, you log and return, here you log and raise exception.  I suspect you want to raise an exception in both cases.

File: cloudinit/distros/netbsd.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+            self.lock_passwd(name)
+
+        # Configure sudo access
+        if 'sudo' in kwargs and kwargs['sudo'] is not False:
+            self.write_sudo_rules(name, kwargs['sudo'])
+
+        # Import SSH keys
+        if 'ssh_authorized_keys' in kwargs:
+            keys = set(kwargs['ssh_authorized_keys']) or []
+            ssh_util.setup_user_keys(keys, name, options=None)
```
Comment (raharper) [inline cloudinit/distros/netbsd.py]: No need to pass options=None; that's the default in the signature.

File: cloudinit/distros/netbsd.py [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+
+    def generate_fallback_config(self):
+        nconf = {'config': [], 'version': 1}
+        for mac, name in net.get_interfaces_by_mac().items():
+            nconf['config'].append(
+                {'type': 'physical', 'name': name,
+                 'mac_address': mac, 'subnets': [{'type': 'dhcp'}]})
+        return nconf
```
Comment (raharper) [inline cloudinit/distros/netbsd.py]: Could you comment here why you want to dhcp on all interfaces versus picking one?

File: cloudinit/distros/netbsd.py [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+
+    def _write_network_config(self, netconfig):
+        return self._supported_write_network_config(netconfig)
+
+    def apply_locale(self, locale, out_fn=None):
+        LOG.debug('Cannot set the locale.')
```
Comment (raharper) [inline cloudinit/distros/netbsd.py]: Is this really true?  It might be worth at least showing the locale that is set if you cannot set it.

File: cloudinit/distros/openbsd.py [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+            cmd = ['pkg_add', '-U']
+        elif command == 'remove':
+            cmd = ['pkg_delete']
+        if args and isinstance(args, str):
+            cmd.append(args)
+        elif args and isinstance(args, list):
+            cmd.extend(args)
+
+        pkglist = util.expand_package_list('%s-%s', pkgs)
+        cmd.extend(pkglist)
```
Comment (raharper) [inline cloudinit/distros/openbsd.py]: if command is not in ['install', 'remove'] then cmd is undefined, so you can't cmd.append or cmd.extend.

File: cloudinit/distros/openbsd.py [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+        return
+
+    def install_packages(self, pkglist):
+        self.update_package_sources()
+        self.package_command('install', pkgs=pkglist)
+
+    def package_command(self, command, args=None, pkgs=None):
+        if pkgs is None:
+            pkgs = []
+
```
Comment (raharper) [inline cloudinit/distros/openbsd.py]:
```suggestion
        if not command or command not in ['install', 'remove']:
            LOG.debug('Invalid package command: %s', command)
            return    
```

File: cloudinit/net/bsd.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+                            'Interface renaming is '
+                            'not supported on this OS'))
+                        device_name = cur_name
+
+            else:
+                device_name = ifname_by_mac[device_mac]
+
+            LOG.info('Configuring interface %s', device_name)
+
+            self.interface_configurations[device_name] = 'DHCP'
```
Comment (raharper) [inline cloudinit/net/bsd.py]: Every device uses DHCP?

File: cloudinit/net/bsd.py [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+        bsd_util.set_rc_config_value(key, value, fn=fn)
+
+    def __init__(self, config=None):
+        if not config:
+            config = {}
+        self.target = None
+        self.interface_configurations = {}
+        self._postcmds = config.get('postcmds', True)
+
+    def _ifconfig_entries(self, settings, target=None):
```
Comment (raharper) [inline cloudinit/net/bsd.py]: This method accepts target, but it is not used.

File: cloudinit/net/bsd.py [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+                        continue
+                    LOG.debug('Configuring dev %s with %s / %s', device_name,
+                              subnet.get('address'), subnet.get('netmask'))
+
+                    self.interface_configurations[device_name] = {
+                        'address': subnet.get('address'),
+                        'netmask': subnet.get('netmask'),
+                    }
+
+    def _route_entries(self, settings, target=None):
```
Comment (raharper) [inline cloudinit/net/bsd.py]: This method accepts target but it is not used.

File: cloudinit/net/openbsd.py [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+class Renderer(cloudinit.net.bsd.BSDRenderer):
+
+    def __init__(self, config=None):
+        super(Renderer, self).__init__()
+
+    def write_config(self):
+        for device_name, v in self.interface_configurations.items():
+            if_file = 'etc/hostname.%s' % device_name
+            fn = util.target_path(self.target, if_file)
+            with open(fn, 'w') as fd:
```
Comment (raharper) [inline cloudinit/net/openbsd.py]: util.write_files()

File: cloudinit/net/bsd.py [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+            except ValueError:
+                util.logexc(LOG, "Failed to add search domain %s", domain)
+        util.write_file(
+            util.target_path(target, self.resolv_conf_fn),
+            str(resolvconf), 0o644)
+
+    def render_network_state(self, network_state, templates=None, target=None):
+        self._ifconfig_entries(settings=network_state)
+        self._route_entries(settings=network_state)
+        self._resolve_conf(settings=network_state)
```
Comment (raharper) [inline cloudinit/net/bsd.py]: target is not passed to _resolv_conf which does use its target parameter.

File: cloudinit/net/bsd.py [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+
+        self.write_config()
```
Comment (raharper) [inline cloudinit/net/bsd.py]: write_config should accept target since it will be writing files and render_network_state is the entry point where target gets passed from the distro class.

File: cloudinit/net/bsd.py [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+        self.start_services(run=self._postcmds)
+
+    def dhcp_interfaces(self):
+        ic = self.interface_configurations.items
+        return [k for k, v in ic() if v == 'DHCP']
```
Comment (raharper) [inline cloudinit/net/bsd.py]:
```suggestion
        return [k for k, v in self.interface_configurations.items() if v == 'DHCP']
```

File: cloudinit/net/freebsd.py [hunk 21]
```diff
@@ -1,156 +1,29 @@
...
-            if network == '0.0.0.0':
-                self._update_rc_conf(
-                    {'defaultrouter': gateway}, target=target)
+        self._route_cpt = 0
+        super(Renderer, self).__init__()
+
+    def rename_interface(self, cur_name, device_name):
+        self.set_rc_config_value('ifconfig_%s_name' % cur_name, device_name)
+
+    def write_config(self):
```
Comment (raharper) [inline cloudinit/net/freebsd.py]:
```suggestion
    def write_config(self, target=None):
```

File: cloudinit/net/openbsd.py [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+                    fd.write(v.get('address') + ' netmask '
+                             + v.get('netmask') + '\n')
+                else:
+                    fd.write('dhcp')
+
+    def start_services(self, run=False):
+        if not self._postcmds:
+            LOG.debug("netbsd generate postcmd disabled")
+            return
+        util.subp(['sh', '/etc/netstart'], capture=True)
```
Comment (raharper) [inline cloudinit/net/openbsd.py]: is /etc/netstart not already executable?

File: cloudinit/net/openbsd.py [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+
+    def set_route(self, network, netmask, gateway):
+        if network == '0.0.0.0':
+            if_file = 'etc/mygate'
+            fn = util.target_path(self.target, if_file)
+            with open(fn, 'w') as fd:
```
Comment (raharper) [inline cloudinit/net/openbsd.py]: util.write_file()

File: cloudinit/net/openbsd.py [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+    def start_services(self, run=False):
+        if not self._postcmds:
+            LOG.debug("netbsd generate postcmd disabled")
+            return
+        util.subp(['sh', '/etc/netstart'], capture=True)
+
+    def set_route(self, network, netmask, gateway):
+        if network == '0.0.0.0':
+            if_file = 'etc/mygate'
+            fn = util.target_path(self.target, if_file)
```
Comment (raharper) [inline cloudinit/net/openbsd.py]: Can we refactor writing this file out in write_config(); that way all writes happen in a single location.

File: cloudinit/tests/helpers.py [hunk 25]
```diff
@@ -362,6 +362,7 @@ def reRoot(self, root=None):
             root = self.tmp_dir()
         self.patchUtils(root)
         self.patchOS(root)
+        self.patchOpen(root)
```
Comment (raharper) [inline cloudinit/tests/helpers.py]: This is quite the change, affects all testcases which use FilesystemMockingTestCase users.  Is there a specific reason why this change?

File: cloudinit/sources/DataSourceNoCloud.py [hunk 26]
```diff
@@ -40,6 +40,16 @@ def _get_devices(self, label):
...
+        elif util.is_NetBSD():
+            out, _err = util.subp(['sysctl', '-n', 'hw.disknames'], rcs=[0])
+            devlist = []
+            for dev in out.split():
+                mscdlabel_out, _ = util.subp(['mscdlabel', dev], rcs=[0, 1])
+                if ('label "%s"' % label) in mscdlabel_out:
+                    devlist.append('/dev/' + dev)
+                    devlist.append('/dev/' + dev + 'a')  # NetBSD 7
+        elif util.is_OpenBSD():
+            devlist = ['/dev/cd0a']
```
Comment (raharper) [inline cloudinit/sources/DataSourceNoCloud.py]: why this change if you've fixed util.find_devs_with() for NetBSD and OpenBSD ?

File: setup.py [hunk 27]
```diff
@@ -152,6 +155,8 @@ def render_tmpl(template, mode=None):
 INITSYS_ROOTS = {
     'sysvinit': 'etc/rc.d/init.d',
     'sysvinit_freebsd': 'usr/local/etc/rc.d',
+    'sysvinit_netbsd': 'usr/local/etc/rc.d',
+    'sysvinit_openbsd': 'tmp',
```
Comment (raharper) [inline setup.py]: tmp really?

File: cloudinit/distros/netbsd.py (hunk 6 above)
Comment (igalic) [reply cloudinit/distros/netbsd.py]: i think @OddBloke suggested elsewhere that we have a variable for these commands, so we don't have to override the entire function, or clutter base with is_x()

**Comment by OddBloke** [comment]
This requires reconciliation with master, so marking it as `wip` for now.

**Comment by goneri** [comment]
@OddBloke Could you please remove the `wip` label? the Pr is ready for review.

File: cloudinit/net/openbsd.py [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+from cloudinit import util
+import cloudinit.net.bsd
+
+LOG = logging.getLogger(__name__)
+
+
+class Renderer(cloudinit.net.bsd.BSDRenderer):
+
+    def __init__(self, config=None):
+        super(Renderer, self).__init__()
```
Comment (OddBloke) [inline cloudinit/net/openbsd.py]: This definition doesn't do anything, could it be omitted?

File: cloudinit/distros/openbsd.py (hunk 12 above)
Comment (OddBloke) [reply cloudinit/distros/openbsd.py]: This still needs to be resolved.

File: cloudinit/distros/openbsd.py [hunk 29]
```diff
@@ -0,0 +1,150 @@
...
+            return fd.read()
+
+    def _write_hostname(self, hostname, filename):
+        content = hostname + '\n'
+        util.write_file(self.hostname_conf_fn, content)
+
+    def _get_add_member_to_group_cmd(self, member_name, group_name):
+        return ['usermod', '-G', group_name, member_name]
+
+    def add_user(self, name, **kwargs):
```
[the remark this answers is not in the document]
Comment (goneri) [reply cloudinit/distros/openbsd.py]: Yeah. I agree. OpenBSD will now inherit from NetBSD.

File: cloudinit/distros/openbsd.py [hunk 30]
```diff
@@ -0,0 +1,48 @@
...
+import platform
+
+import cloudinit.distros.netbsd
+from cloudinit import log as logging
+from cloudinit import util
+
+LOG = logging.getLogger(__name__)
+
+
+class Distro(cloudinit.distros.netbsd.NetBSD):
```
Comment (goneri) [inline cloudinit/distros/openbsd.py]: @OddBloke the class now inherite from the NetBSD class.

File: cloudinit/distros/netbsd.py [hunk 31]
```diff
@@ -13,7 +13,7 @@
 LOG = logging.getLogger(__name__)
 
 
-class Distro(cloudinit.distros.bsd.BSD):
+class NetBSD(cloudinit.distros.bsd.BSD):
```
Comment (OddBloke) [inline cloudinit/distros/netbsd.py]: Given that the naming doesn't indicate this is used for OpenBSD too, I'd like to see a comment/docstring here that indicates that it is.

File: cloudinit/net/openbsd.py [hunk 32]
```diff
@@ -0,0 +1,38 @@
...
+import cloudinit.net.bsd
+
+LOG = logging.getLogger(__name__)
+
+
+class Renderer(cloudinit.net.bsd.BSDRenderer):
+
+    def write_config(self):
+        for device_name, v in self.interface_configurations.items():
+            if_file = 'etc/hostname.%s' % device_name
```
Comment (OddBloke) [inline cloudinit/net/openbsd.py]:
```suggestion
            if_file = 'etc/hostname.{}'.format(device_name)
```

File: cloudinit/sources/DataSourceConfigDrive.py [hunk 33]
```diff
@@ -72,7 +72,7 @@ def _get_data(self):
             dslist = self.sys_cfg.get('datasource_list')
             for dev in find_candidate_devs(dslist=dslist):
                 mtype = None
-                if (util.is_FreeBSD() or util.is_NetBSD()):
+                if (util.is_FreeBSD() or util.is_NetBSD() or util.is_NetBSD()):
```
Comment (OddBloke) [inline cloudinit/sources/DataSourceConfigDrive.py]: Once #268 lands, this should be `util.is_BSD`.

File: tools/build-on-openbsd [hunk 34]
```diff
@@ -0,0 +1,27 @@
...
+   py3-setuptools
+   py3-six
+   py3-yaml
+   sudo--
+"
+[ -f "$depschecked" ] || pkg_add ${pkgs} || fail "install packages"
+
+touch $depschecked
+
+python3.7 setup.py build
```
Comment (OddBloke) [inline tools/build-on-openbsd]: Would we save ourselves some future work if we just used `python3` here? Or do we need to specify the exact Python version for some reason?

File: cloudinit/util.py [hunk 35]
```diff
@@ -1268,6 +1273,11 @@ def find_devs_with_netbsd(criteria=None, oformat='device',
     return []
 
 
+def find_devs_with_openbsd(criteria=None, oformat='device',
+                           tag=None, no_cache=False, path=None):
+    return ["/dev/cd0a"]
```
Comment (OddBloke) [inline cloudinit/util.py]: Should there at least be an existence check for this instead of returning it unconditionally? (If not, then an explanatory comment in here would be good, I think.)

File: cloudinit/sources/DataSourceNoCloud.py [hunk 36]
```diff
@@ -48,6 +48,15 @@ def _get_devices(self, label):
                 if ('label "%s"' % label) in mscdlabel_out:
                     devlist.append('/dev/' + dev)
                     devlist.append('/dev/' + dev + 'a')  # NetBSD 7
+        elif util.is_OpenBSD():
+            out, _err = util.subp(['sysctl', '-n', 'hw.disknames'], rcs=[0])
+            devlist = []
+            for i in out.split(','):
+                if i[-1] != ':':
```
Comment (OddBloke) [inline cloudinit/sources/DataSourceNoCloud.py]:
```suggestion
                if i.endswith(':'):
```

File: cloudinit/sources/DataSourceNoCloud.py [hunk 37]
```diff
@@ -48,6 +48,15 @@ def _get_devices(self, label):
                 if ('label "%s"' % label) in mscdlabel_out:
                     devlist.append('/dev/' + dev)
                     devlist.append('/dev/' + dev + 'a')  # NetBSD 7
+        elif util.is_OpenBSD():
+            out, _err = util.subp(['sysctl', '-n', 'hw.disknames'], rcs=[0])
+            devlist = []
+            for i in out.split(','):
```
Comment (OddBloke) [inline cloudinit/sources/DataSourceNoCloud.py]: Would be good to use a more descriptive variable name than `i` here; I cannot infer what values it might take currently, and I don't have an OpenBSD system to check for myself.

File: cloudinit/net/openbsd.py (hunk 22 above)
Comment (OddBloke) [reply cloudinit/net/openbsd.py]: This still needs a response.

File: cloudinit/distros/openbsd.py (hunk 30 above)
Comment (OddBloke) [reply cloudinit/distros/openbsd.py]: Thanks, this looks a lot better!

File: cloudinit/distros/openbsd.py [hunk 38]
```diff
@@ -0,0 +1,48 @@
...
+    hostname_conf_fn = '/etc/myname'
+
+    def _read_hostname(self, filename, default=None):
+        with open(self.hostname_conf_fn, 'r') as fd:
+            return fd.read()
```
Comment (OddBloke) [inline cloudinit/distros/openbsd.py]: `util.load_file`?

File: cloudinit/net/openbsd.py [hunk 39]
```diff
@@ -0,0 +1,38 @@
...
+            fn = util.target_path(self.target, if_file)
+            content = 'dhcp\n'
+            if isinstance(v, dict):
+                content = "inet {address} {netmask}\n".format(
+                            address=v.get('address'),
```
Comment (OddBloke) [inline cloudinit/net/openbsd.py]: If both of these aren't present, do we really want to produce "inet None None\n"?

File: tools/build-on-openbsd (hunk 34 above)
Comment (goneri) [reply tools/build-on-openbsd]: Yes, good point.

File: cloudinit/util.py (hunk 35 above)
Comment (goneri) [reply cloudinit/util.py]: Yeah, actually I should reuse the logic from cloudinit/sources/DataSourceNoCloud.py and add some test coverage. I will do that later.

File: cloudinit/util.py [hunk 40]
```diff
@@ -1279,6 +1284,26 @@ def find_devs_with_netbsd(criteria=None, oformat='device',
     return []
 
 
+def find_devs_with_openbsd(criteria=None, oformat='device',
+                           tag=None, no_cache=False, path=None):
+    out, _err = subp(['sysctl', '-n', 'hw.disknames'], rcs=[0])
```
Comment (goneri) [inline cloudinit/util.py]: This time, the function comes with a unit-test, and is in cloudinit.util. This way, I don't need to patch `cloudinit/sources/DataSourceNoCloud.py`.

File: cloudinit/distros/openbsd.py (hunk 30 above)
Comment (goneri) [reply cloudinit/distros/openbsd.py]: Great :-)

File: cloudinit/net/openbsd.py (hunk 22 above)
Comment (goneri) [reply cloudinit/net/openbsd.py]: Nope, it's not:

```
openbsd-66$ ls -l /etc/netstart                                                                                                                                                              
-rw-r--r--  1 root  wheel  9580 Oct 12 10:34 /etc/netstart
```

File: cloudinit/sources/DataSourceNoCloud.py (hunk 37 above)
Comment (goneri) [reply cloudinit/sources/DataSourceNoCloud.py]: the code is now in cloudinit/util.py and has an unit-test.

File: cloudinit/util.py (hunk 35 above)
Comment (goneri) [reply cloudinit/util.py]: done

File: cloudinit/distros/netbsd.py [hunk 41]
```diff
@@ -13,7 +13,9 @@
...
+    """OpenBSD inherite from this class."""
```
Comment (OddBloke) [inline cloudinit/distros/netbsd.py]:
```suggestion
    """
    Distro subclass for NetBSD.
    
    (N.B. OpenBSD inherits from this class.)
    """
```

File: cloudinit/net/__init__.py [hunk 42]
```diff
@@ -857,6 +861,21 @@ def get_interfaces_by_mac_on_netbsd():
     return ret
 
 
+def get_interfaces_by_mac_on_openbsd():
+    ret = {}
+    re_field_match = (
+            r"(?P<ifname>\w+).*lladdr\s"
+            r"(?P<mac>([\da-f]{2}[:-]){5}([\da-f]{2})).*")
```
Comment (OddBloke) [inline cloudinit/net/__init__.py]:
```suggestion
        r"(?P<ifname>\w+).*lladdr\s"
        r"(?P<mac>([\da-f]{2}[:-]){5}([\da-f]{2})).*")
```

File: cloudinit/net/openbsd.py [hunk 43]
```diff
@@ -0,0 +1,41 @@
...
+            if_file = 'etc/hostname.{}'.format(device_name)
+            fn = util.target_path(self.target, if_file)
+            content = 'dhcp\n'
+            try:
+                if isinstance(v, dict):
+                    content = "inet {address} {netmask}\n".format(
+                                address=v['address'],
+                                netmask=v['netmask'])
+            except KeyError:
+                LOG.error("Invalid static configuration for %s", device_name)
```
Comment (OddBloke) [inline cloudinit/net/openbsd.py]: Should the log message mention that we're falling back to DHCP?

File: cloudinit/net/__init__.py [hunk 44]
```diff
@@ -336,6 +336,8 @@ def find_fallback_nic(blacklist_drivers=None):
         return find_fallback_nic_on_freebsd(blacklist_drivers)
     elif util.is_NetBSD():
         return find_fallback_nic_on_netbsd(blacklist_drivers)
+    elif util.is_OpenBSD():
+        return find_fallback_nic_on_netbsd(blacklist_drivers)
```
Comment (OddBloke) [inline cloudinit/net/__init__.py]: Any reason not to do this?

```suggestion
    elif util.is_NetBSD() or util.is_OpenBSD():
        return find_fallback_nic_on_netbsd(blacklist_drivers)
```

And should we consider renaming `find_fallback_nic_on_netbsd` to reflect that it is no longer just used for NetBSD?

File: cloudinit/util.py [hunk 45]
```diff
@@ -1279,6 +1284,26 @@ def find_devs_with_netbsd(criteria=None, oformat='device',
...
+    devlist = []
+    for entry in out.split(','):
+        if not entry.endswith(':'):
+            # ffs partition with a serial, not a config-drive
+            continue
+        if entry == 'fd0:':
+            continue
+        part_id = 'a' if entry.startswith('cd') else 'i'
+        devlist.append(entry[:-1] + part_id)
+    if criteria:
```
Comment (OddBloke) [inline cloudinit/util.py]: No need for this `if`; `None != "TYPE=iso9660"` and `None != "LABEL=CONFIG-2"`, so the nested ifs will cover this case.

File: cloudinit/util.py [hunk 46]
```diff
@@ -1279,6 +1284,26 @@ def find_devs_with_netbsd(criteria=None, oformat='device',
...
+        if criteria == "TYPE=iso9660":
+            devlist = [i for i in devlist if i.startswith('cd')]
+        elif criteria == "LABEL=CONFIG-2":
+            devlist = [i for i in devlist if not i.startswith('cd')]
```
Comment (OddBloke) [inline cloudinit/util.py]: If we see `criteria` that we don't support, should we do something other than ignore them? (I'm not sure what the expected behaviour here is, this isn't a leading question.)

File: cloudinit/util.py (hunk 40 above)
Comment (OddBloke) [reply cloudinit/util.py]: Thanks for this!

File: cloudinit/net/openbsd.py (hunk 43 above)
Comment (goneri) [reply cloudinit/net/openbsd.py]: I thought about it too, but I think it will obvious enough.

File: cloudinit/net/__init__.py (hunk 44 above)
Comment (goneri) [reply cloudinit/net/__init__.py]: I don't have a much better name, do you? Ultimately, I would like to move the functions in cloudinit.distros.netbsd.

File: cloudinit/util.py (hunk 45 above)
Comment (goneri) [reply cloudinit/util.py]: indeed. thanks.

File: cloudinit/util.py (hunk 46 above)
Comment (goneri) [reply cloudinit/util.py]: Ok, I added an extra `if` case of that.

File: cloudinit/net/__init__.py (hunk 44 above)
Comment (OddBloke) [reply cloudinit/net/__init__.py]: I was really just thinking something like `find_fallback_nic_on_netbsd_or_openbsd` to give future developers (who might only be considering NetBSD) a breadcrumb to realise that the function isn't NetBSD-specific.

File: cloudinit/net/openbsd.py (hunk 43 above)
Comment (OddBloke) [reply cloudinit/net/openbsd.py]: I think it will be obvious that a system is using DHCP; I'm not so sure it will be obvious _why_ if people can't search "dhcp" in the log and find this line.

File: cloudinit/distros/openbsd.py [hunk 47]
```diff
@@ -0,0 +1,47 @@
...
+            util.logexc(LOG, "Failed to lock user %s", name)
+            raise
+
+    def _get_pkg_cmd_environ(self):
+        """Return env vars used in OpenBSD package_command operations"""
+        os_release = platform.release()
+        os_arch = platform.machine()
+        e = os.environ.copy()
+        e['PKG_PATH'] = (
+                'ftp://ftp.openbsd.org/pub/OpenBSD/{os_release}/'
```
Comment (raharper) [inline cloudinit/distros/openbsd.py]: Does the NetBSD Distro class also use the OpenBSD ftp resource?

File: cloudinit/net/__init__.py [hunk 48]
```diff
@@ -335,12 +335,14 @@ def find_fallback_nic(blacklist_drivers=None):
     if util.is_FreeBSD():
         return find_fallback_nic_on_freebsd(blacklist_drivers)
     elif util.is_NetBSD():
-        return find_fallback_nic_on_netbsd(blacklist_drivers)
+        return find_fallback_nic_on_netbsd_or_openbsd(blacklist_drivers)
+    elif util.is_OpenBSD():
+        return find_fallback_nic_on_netbsd_or_openbsd(blacklist_drivers)
```
Comment (raharper) [inline cloudinit/net/__init__.py]:
```
elif util.is_NetBSD() or util.is_OpenBSD():
    return find_fallback_nic_on_netbsd_or_openbsd(blacklist_drivers)
```

File: cloudinit/distros/openbsd.py (hunk 47 above)
Comment (goneri) [reply cloudinit/distros/openbsd.py]: well, no. It uses a NetBSD URL. I'm not sure I understand the question.

Comment (raharper) [reply cloudinit/distros/openbsd.py]: Sorry, I was confused, the OpenBSD class is subclassed from NetBSD and then I was thinking this was the NetBSD path.

File: cloudinit/net/openbsd.py [hunk 49]
```diff
@@ -0,0 +1,43 @@
...
+from cloudinit import util
+import cloudinit.net.bsd
+
+LOG = logging.getLogger(__name__)
+
+
+class Renderer(cloudinit.net.bsd.BSDRenderer):
+
+    def write_config(self):
+        for device_name, v in self.interface_configurations.items():
```
Comment (raharper) [inline cloudinit/net/openbsd.py]: Is it typical to DHCP on all interfaces?  In clouds multi-nic DHCP will cause issues with default routing (if the DHCP server sends a gateway/route for each interface.  Certainly on Azure, secondary nics who DHCP will break networking going out of the box or hitting the IMDS due to traffic not flowing out of the primary NIC.

I strongly discourage dhcp'ing on all interfaces without a way to set a device or route metric.

File: cloudinit/net/openbsd.py [hunk 50]
```diff
@@ -0,0 +1,43 @@
...
+                                netmask=v['netmask'])
+            except KeyError:
+                LOG.error((
+                    "Invalid static configuration for %s,"
+                    "falling back to dhcp"), device_name)
+            util.write_file(fn, content)
+
+    def start_services(self, run=False):
+        if not self._postcmds:
+            LOG.debug("netbsd generate postcmd disabled")
```
Comment (raharper) [inline cloudinit/net/openbsd.py]: s/netbsd/openbsd ?

File: cloudinit/net/openbsd.py (hunk 49 above)
Comment (goneri) [reply cloudinit/net/openbsd.py]: ok, I see. 

> I'd like to clarify expectations on dhcp on all nics, and one name fix up (net -> open).

Done, we don't default on DHCP anymore.
